### PR TITLE
Use crew environment directories

### PR DESCRIPTION
**Pull request by Zopolis4**
Also update fish to source from there.

### Run the following to get this pull request's changes locally for testing.
```
CREW_TESTING_REPO=https://github.com/Zopolis4/chromebrew.git CREW_TESTING_BRANCH=scopedin CREW_TESTING=1 crew update
```

### CODE REVIEW

File: packages/bluefish.rb [hunk 1]
```diff
@@ -47,7 +47,7 @@ def self.postinstall
     puts
     puts 'To complete the installation, execute the following:'.lightblue
     puts "update-mime-database #{CREW_PREFIX}/share/mime".lightblue
-    puts "echo 'alias bluefish=\"WAYLAND_DISPLAY=wayland-0 DISPLAY=\'\' GDK_BACKEND=wayland bluefish\"' >> ~/.bashrc".lightblue
+    puts "echo 'alias bluefish=\"WAYLAND_DISPLAY=wayland-0 DISPLAY=\'\' GDK_BACKEND=wayland bluefish\"' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
```
Comment (uberhacker) [inline packages/bluefish.rb:50]: Actually, why don't we remove this line from the postinstall section and add it to the package instead to `#{CREW_DEST_PREFIX}/etc/env.d/bluefish`.  That way we don't have to append the main profile which could potentially end up with duplicate lines and the lines are removed when the package is removed.

File: packages/broadway.rb [hunk 2]
```diff
@@ -58,11 +58,11 @@ def self.install
...
-    puts "echo '# Broadway environment variables + daemon' >> ~/.bashrc".lightblue
-    puts "echo '# See https://developer.gnome.org/gtk3/stable/gtk-broadway.html' >> ~/.bashrc".lightblue
-    puts "echo 'alias startbroadway=\"set -a && source ~/.broadway.env && set +a && initbroadway\"' >> ~/.bashrc".lightblue
-    puts "echo 'startbroadway' >> ~/.bashrc".lightblue
-    puts 'source ~/.bashrc'.lightblue
+    puts "echo '# Broadway environment variables + daemon' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
+    puts "echo '# See https://developer.gnome.org/gtk3/stable/gtk-broadway.html' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
+    puts "echo 'alias startbroadway=\"set -a && source ~/.broadway.env && set +a && initbroadway\"' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
+    puts "echo 'startbroadway' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
+    puts "source #{CREW_PREFIX}/etc/env.d/profile".lightblue
```
Comment (uberhacker) [inline packages/broadway.rb:65]: See above.

File: packages/composer.rb [hunk 3]
```diff
@@ -33,6 +33,6 @@ def self.install
   def self.postinstall
     FileUtils.ln_sf "#{CREW_PREFIX}/.config", "#{HOME}/.config"
     puts "\nTo finish the installation, execute the following:".lightblue
-    puts "echo 'export PATH=\$HOME/.config/composer/vendor/bin:\$PATH' >> ~/.bashrc && . ~/.bashrc\n".lightblue
+    puts "echo 'export PATH=\$HOME/.config/composer/vendor/bin:\$PATH' >> #{CREW_PREFIX}/etc/env.d/profile && . #{CREW_PREFIX}/etc/env.d/profile\n".lightblue
```
Comment (uberhacker) [inline packages/composer.rb:36]: See above.

File: packages/cronie.rb [hunk 4]
```diff
@@ -83,11 +83,11 @@ def self.postinstall
...
-    puts "echo '# start the crond daemon' >> ~/.bashrc".lightblue
-    puts "echo 'if [ -f #{CREW_PREFIX}/bin/startcrond ]; then' >> ~/.bashrc".lightblue
-    puts "echo '  #{CREW_PREFIX}/bin/startcrond' >> ~/.bashrc".lightblue
-    puts "echo 'fi' >> ~/.bashrc".lightblue
-    puts 'source ~/.bashrc'.lightblue
+    puts "echo '# start the crond daemon' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
+    puts "echo 'if [ -f #{CREW_PREFIX}/bin/startcrond ]; then' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
+    puts "echo '  #{CREW_PREFIX}/bin/startcrond' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
+    puts "echo 'fi' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
+    puts "source #{CREW_PREFIX}/etc/env.d/profile".lightblue
```
Comment (uberhacker) [inline packages/cronie.rb:90]: See above.

**Comment by uberhacker** [comment]
@Zopolis4: I appreciate your enthusiasm and effort for taking on this major task.

**Review by uberhacker — CHANGES_REQUESTED** [review]
Please see the comment in bluefish.rb.  This general idea should be implemented for all packages, if possible.

File: packages/bluefish.rb (hunk 1 above)
Comment (uberhacker) [reply packages/bluefish.rb:50]: It won't hurt to actually run the `update-mime-database` command without prompting too.

Comment (Zopolis4) [reply packages/bluefish.rb:50]: Is it required to move the writing to `#{CREW_DEST_PREFIX}/etc/env.d/bluefish` out of postinstall? It's still in postinstall in fish.

Comment (uberhacker) [reply packages/bluefish.rb:50]: Yes, it can go in the `def self.build` section instead.

**Review by uberhacker — CHANGES_REQUESTED** [review]
I don't know what happened to the earlier comments but there are several `FileUtils.mkdir_p` directories that include files.  Those should be removed.  I believe @supechicken pointed this out too.

**Comment by Zopolis4** [comment]
Done.

**Comment by Zopolis4** [comment]
I've done all of the previously manual packages, leaving ones that should be up to the user's discretion alone.

File: packages/broadway.rb [hunk 5]
```diff
@@ -16,58 +16,67 @@ class Broadway < Package
...
     system "install -Dm755 stopbroadway #{CREW_DEST_PREFIX}/bin/stopbroadway"
-    system "install -Dm644 .broadway.env #{CREW_DEST_HOME}/.broadway.env"
   end
 
   def self.postinstall
+    FileUtils.mkdir_p "#{CREW_DEST_PREFIX}/etc/env.d"
+    @broadway = <<~BROADWAYEOF
+      alias startbroadway="set -a && source #{CREW_DEST_HOME}/.broadway.env && set +a && initbroadway"
+    BROADWAYEOF
+    File.write("#{CREW_DEST_PREFIX}/etc/env.d/broadway", @broadway)
```
Comment (uberhacker) [inline packages/broadway.rb:70]: Lines 66-70 should be moved to the build section.

File: packages/broadway.rb [hunk 6]
```diff
@@ -16,58 +16,67 @@ class Broadway < Package
...
+    system "source #{CREW_PREFIX}/etc/env.d/profile"
```
Comment (uberhacker) [inline packages/broadway.rb:71]: This won't work and should be removed.

File: packages/broadway.rb [hunk 7]
```diff
@@ -16,58 +16,67 @@ class Broadway < Package
...
+      fi
+      BROADWAYD=$(pidof broadwayd 2>/dev/null)
+      if [ ! -z "${BROADWAYD}" ]; then
+        echo "broadwayd process ${BROADWAYD} is running"
+      else
+        echo "broadwayd failed to start"
+        exit 1
+      fi
+    INITBROADWAYEOF
+    File.write('initbroadway', @initbroadway)
```
Comment (uberhacker) [inline packages/broadway.rb:41]:
```
FileUtils.mkdir_p "#{CREW_DEST_PREFIX}/bin"
File.write("#{CREW_DEST_PREFIX}/bin/initbroadway", @initbroadway)
```

File: packages/broadway.rb [hunk 8]
```diff
@@ -16,58 +16,67 @@ class Broadway < Package
...
+      fi
+      BROADWAYD=$(pidof broadwayd 2> /dev/null)
+      if [ -z "${BROADWAYD}" ]; then
+        echo "broadwayd process stopped"
+      else
+        echo "broadwayd process ${BROADWAYD} is running"
+        exit 1
+      fi
+    STOPBROADWAYEOF
+    File.write('stopbroadway', @stopbroadway)
```
Comment (uberhacker) [inline packages/broadway.rb:57]:
```
File.write("#{CREW_DEST_PREFIX}/bin/stopbroadway", @initbroadway)
```

File: packages/broadway.rb [hunk 9]
```diff
@@ -16,58 +16,67 @@ class Broadway < Package
...
-    puts "echo 'startbroadway' >> ~/.bashrc".lightblue
-    puts 'source ~/.bashrc'.lightblue
-    puts
+    puts "echo '# Broadway environment variables + daemon' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
+    puts "echo '# See https://developer.gnome.org/gtk3/stable/gtk-broadway.html' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
     puts "To start the broadwayd daemon, run 'startbroadway'".lightblue
     puts "To stop the broadwayd daemon, run 'stopbroadway'".lightblue
     puts
-    puts 'To adjust environment variables, edit ~/.broadway.env'.lightblue
+    puts "To adjust environment variables, edit #{CREW_DEST_HOME}/.broadway.env".lightblue
```
Comment (uberhacker) [inline packages/broadway.rb:79]: Good change.  Always best to abstract directories when possible.

File: packages/composer.rb [hunk 10]
```diff
@@ -28,11 +28,14 @@ def self.install
...
+    @composer = <<~COMPOSEREOF
+      PATH=$HOME/.config/composer/vendor/bin:$PATH
+    COMPOSEREOF
+    File.write("#{CREW_DEST_PREFIX}/etc/env.d/composer", @composer)
   end
 
   def self.postinstall
     FileUtils.ln_sf "#{CREW_PREFIX}/.config", "#{HOME}/.config"
-    puts "\nTo finish the installation, execute the following:".lightblue
-    puts "echo 'export PATH=\$HOME/.config/composer/vendor/bin:\$PATH' >> ~/.bashrc && . ~/.bashrc\n".lightblue
```
Comment (uberhacker) [inline packages/composer.rb:36]: It would be good to keep this:
```
    puts "\nTo finish the installation, execute the following:".lightblue
    puts "source ~/.bashrc\n".lightblue
```

File: packages/cronie.rb [hunk 11]
```diff
@@ -30,34 +30,42 @@ def self.build
...
+      fi
+      CRON=$(pidof $CROND 2> /dev/null)
+      if [ ! -z "$CRON" ]; then
+        echo "crond process $CRON is running"
+      else
+        echo "crond failed to start"
+        exit 1
+      fi
+    STARTCRONDEOF
+    File.write('startcrond', @startcrond)
```
Comment (uberhacker) [inline packages/cronie.rb:51]: Write to `#{CREW_DEST_PREFIX}/bin/startcrond`?

File: packages/cronie.rb [hunk 12]
```diff
@@ -30,34 +30,42 @@ def self.build
...
+      fi
+      CRON=$(pidof $CROND 2> /dev/null)
+      if [ -z "$CRON" ]; then
+        echo "crond process stopped"
+      else
+        echo "crond process $CRON is running"
+        exit 1
+      fi
+    STOPCRONDEOF
+    File.write('stopcrond', @stopcrond)
```
Comment (uberhacker) [inline packages/cronie.rb:68]: Write to `#{CREW_DEST_PREFIX}/bin/stopcrond`?

File: packages/cronie.rb [hunk 13]
```diff
@@ -83,11 +91,11 @@ def self.postinstall
...
     puts 'To start the crond daemon at login, execute the following:'.lightblue
-    puts "echo '# start the crond daemon' >> ~/.bashrc".lightblue
-    puts "echo 'if [ -f #{CREW_PREFIX}/bin/startcrond ]; then' >> ~/.bashrc".lightblue
-    puts "echo '  #{CREW_PREFIX}/bin/startcrond' >> ~/.bashrc".lightblue
-    puts "echo 'fi' >> ~/.bashrc".lightblue
-    puts 'source ~/.bashrc'.lightblue
+    puts "echo '# start the crond daemon' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
+    puts "echo 'if [ -f #{CREW_PREFIX}/bin/startcrond ]; then' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
+    puts "echo '  #{CREW_PREFIX}/bin/startcrond' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
+    puts "echo 'fi' >> #{CREW_PREFIX}/etc/env.d/profile".lightblue
```
Comment (uberhacker) [inline packages/cronie.rb:97]: Write to `#{CREW_DEST_PREFIX}/etc/env.d/cronie`?

File: packages/cronie.rb [hunk 14]
```diff
@@ -83,11 +91,11 @@ def self.postinstall
...
+    puts "source #{CREW_PREFIX}/etc/env.d/profile".lightblue
```
Comment (uberhacker) [inline packages/cronie.rb:98]: This won't work and should be removed.  Prompt to `source ~/.bashrc` instead?

File: packages/gcloud.rb [hunk 15]
```diff
@@ -52,7 +52,7 @@ def self.install
   def self.postinstall
     puts
     puts 'To finish the installation, execute the following:'.lightblue
-    puts 'source ~/.bashrc && gcloud init'.lightblue
+    puts "source #{CREW_PREFIX}/etc/env.d/profile && gcloud init".lightblue
```
Comment (uberhacker) [inline packages/gcloud.rb:55]: This won't work and should be removed.  Prompt to `source ~/.bashrc` instead?

File: packages/git_prompt.rb [hunk 16]
```diff
@@ -32,7 +32,7 @@ def self.install
 
   def self.postinstall
     puts "\nTo finish the installation, execute the following:".lightblue
-    puts 'source ~/.bashrc'.lightblue
+    puts "source #{CREW_PREFIX}/etc/env.d/profile".lightblue
```
Comment (uberhacker) [inline packages/git_prompt.rb:35]: This won't work and should be removed.  Prompt to `source ~/.bashrc` instead?  IOW, do not change.

File: packages/handbrake.rb [hunk 17]
```diff
@@ -50,8 +54,5 @@ def self.postinstall
     puts
     puts "Type 'HandBrakeCLI' for the command line.".lightblue
     puts
-    puts 'To complete the installation, execute the following:'.lightblue
-    puts "echo 'alias ghb=\"WAYLAND_DISPLAY=wayland-0 DISPLAY=\'\' GDK_BACKEND=wayland ghb\"' >> ~/.bashrc".lightblue
-    puts
```
Comment (uberhacker) [inline packages/handbrake.rb:55]: Prompt to `source ~/.bashrc` in postinstall?

File: packages/homebank.rb [hunk 18]
```diff
@@ -33,15 +33,13 @@ class Homebank < Package
...
   end
 
   def self.install
     system 'make', "DESTDIR=#{CREW_DEST_DIR}", 'install'
   end
-
-  def self.postinstall
-    puts
-    puts 'To complete the installation, execute the following:'.lightblue
-    puts "echo 'alias homebank=\"WAYLAND_DISPLAY=wayland-0 DISPLAY=\'\' GDK_BACKEND=wayland homebank\"' >> ~/.bashrc".lightblue
```
Comment (uberhacker) [inline packages/homebank.rb:45]: Prompt to `source ~/.bashrc`?

**Review by uberhacker — CHANGES_REQUESTED** [review]
Please see comments.

File: packages/broadway.rb (hunk 8 above)
Comment (Zopolis4) [reply packages/broadway.rb:57]: See the install section. (This applies to `cronie.rb` as well.)

File: packages/cronie.rb (hunk 14 above)
Comment (Zopolis4) [reply packages/cronie.rb:98]: Would it not make more sense to source `~/.profile`? I'm a little confused about how exactly the crew profile stuff works.

**Comment by Zopolis4** [comment]
Given that this PR encompasses 3 types of changes, I'm going to be splitting it into 3 different prs.

**Comment by Zopolis4** [comment]
tl;dr-- chromebrew is good, our setup for gui packages is less good, and our setup for service/daemon packages is even less good. I'll try to make our setup for gui packages more good, and try to do the same for our service/daemon packages, although that'll be tougher. 

after those two have been fixed i might look into the remaining packages using `~/.bashrc`, see how many of them actually should even be going near `~/.bashrc`, fix the ones that dont need to be doing it, then maybe get the remaining packages to do something more in line with `crew_profile_base`.

**Comment by Zopolis4** [comment]
closing because the changes in this pr are not what i'm going for anymore
